sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Helpers/RegistryArtifact.cs

```diff
@@ -0,0 +1,418 @@
+            _tagOrDigest = tagOrDigest;
+            _registryUri = registryUri;
+            _fullyQualifiedName = $"{registryUri.Host}/{repositoryName}{(IsDigest(tagOrDigest) ? '@' : ':')}{tagOrDigest}";
+
+            _clientDiagnostics = clientDiagnostics;
+            _restClient = restClient;
+        }
+
+        /// <summary> Initializes a new instance of RegistryArtifact for mocking. </summary>
+        protected RegistryArtifact()
+        {
+        }
+
+        #region Registry Artifact/Manifest methods
+
+        /// <summary> Get registry artifact properties by tag or digest. </summary>
+        /// <param name="cancellationToken"> The cancellation token to use. </param>
+        public virtual async Task<Response<ArtifactManifestProperties>> GetManifestPropertiesAsync(CancellationToken cancellationToken = default)
+        {
+            using DiagnosticScope scope = _clientDiagnostics.CreateScope($"{nameof(RegistryArtifact)}.{nameof(GetManifestProperties)}");
+            scope.Start();
+
+            try
+            {
+                string digest = await GetDigestAsync(cancellationToken).ConfigureAwait(false);
+                return await _restClient.GetManifestPropertiesAsync(_repositoryName, digest, cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception e)
+            {
+                scope.Failed(e);
+                throw;
+            }
+        }
+
+        /// <summary> Get registry artifact properties by tag or digest. </summary>
+        /// <param name="cancellationToken"> The cancellation token to use. </param>
+        public virtual Response<ArtifactManifestProperties> GetManifestProperties(CancellationToken cancellationToken = default)
+        {
+            using DiagnosticScope scope = _clientDiagnostics.CreateScope($"{nameof(RegistryArtifact)}.{nameof(GetManifestProperties)}");
+            scope.Start();
+
+            try
+            {
+                string digest = GetDigest(cancellationToken);
+                return _restClient.GetManifestProperties(_repositoryName, digest, cancellationToken);
+            }
+            catch (Exception e)
+            {
+                scope.Failed(e);
+                throw;
+            }
+        }
+
+        private async Task<string> GetDigestAsync(CancellationToken cancellationToken)
+        {
+            _digest ??= IsDigest(_tagOrDigest) ? _tagOrDigest :
+                (await _restClient.GetTagPropertiesAsync(_repositoryName, _tagOrDigest, cancellationToken).ConfigureAwait(false)).Value.Digest;
+
+            return _digest;
+        }
+
+        private string GetDigest(CancellationToken cancellationToken)
+        {
+            _digest ??= IsDigest(_tagOrDigest) ? _tagOrDigest :
+                _restClient.GetTagProperties(_repositoryName, _tagOrDigest, cancellationToken).Value.Digest;
+
+            return _digest;
+        }
+
+        private static bool IsDigest(string tagOrDigest)
```

Comment:
Not sure about the service, but in the docker world a digest will have a ":" as well, e.g. "sha256:2ad7b012c9d05a8b65337ce1591e9517fa7f0305753885dfad06261e52c3027b". This is standard nomenclature. Does the service not return this format then? Of the service, this doesn't seem forward forward compatible and should probably be returned in results eventually if, for example, the world moves to SHA384 or SHA512 manifest IDs - perhaps sometime before the heat-death of the universe. :)

Reply:
We're using knowledge of the [Docker Content Digest](https://docs.docker.com/registry/spec/api/#content-digests) here and knowledge that ':' is an illegal character in an image tag to differentiate between digest and tag here.  Per the docker content digest grammar, this should accommodate the use of different algorithms to compute the digest.